includes/class-wc-stripe-api.php

```diff
@@ -250,6 +278,80 @@ public static function retrieve( $api ) {
 		return json_decode( $response['body'] );
 	}
 
+	/**
+	 * Checks if the Stripe API for the current mode (i.e. test or live) is rate limited.
+	 *
+	 * @return bool True if the Stripe API is rate limited, false otherwise.
+	 */
+	public static function is_stripe_api_rate_limited() {
+		$rate_limit_option_key = WC_Stripe_Mode::is_test() ? self::TEST_MODE_STRIPE_API_RATE_LIMIT_OPTION_KEY : self::LIVE_MODE_STRIPE_API_RATE_LIMIT_OPTION_KEY;
+
+		$rate_limit_expiration = get_option( $rate_limit_option_key );
+		if ( ! $rate_limit_expiration ) {
+			return false;
+		}
+
+		$now = time();
+		if ( $now > $rate_limit_expiration ) {
+			delete_option( $rate_limit_option_key );
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Helper function to check error responses from Stripe and ensure we prevent unnecessary API calls,
+	 * primarily in cases where we have been rate limited or we don't have valid keys..
+	 *
+	 * @param array|WP_Error $response The response from the Stripe API.
+	 * @return void
+	 */
+	protected static function check_stripe_api_error_response( $response ) {
+		// If we don't have an array for $response, return early, as we won't have an HTTP status code.
+		if ( ! is_array( $response ) ) {
+			return;
+		}
+
+		// We specifically want to check $response['response']['code']. If it's not present, return early.
+		if ( ! isset( $response['response'] ) || ! is_array( $response['response'] ) || ! isset( $response['response']['code'] ) ) {
+			return;
+		}
+
+		$status_code = $response['response']['code'];
+
+		if ( 429 === $status_code ) {
+			// Stripe has rate limited us, so disable API calls for a period of time.
+			$is_test_mode = WC_Stripe_Mode::is_test();
+
+			$timestamp = time();
+			$rate_limit_option_key = $is_test_mode ? self::TEST_MODE_STRIPE_API_RATE_LIMIT_OPTION_KEY : self::LIVE_MODE_STRIPE_API_RATE_LIMIT_OPTION_KEY;
+			update_option( $rate_limit_option_key, $timestamp + self::STRIPE_API_RATE_LIMIT_DURATION );
+
+			$mode = $is_test_mode ? 'test' : 'LIVE';
+			$message = "Stripe {$mode} mode API has been rate limited, disabling API calls for " . self::STRIPE_API_RATE_LIMIT_DURATION . ' seconds.';
+
+			error_log( 'woocommerce-gateway-stripe: WARNING: ' . $message );
```

Comment:
Did we mean to have a separate `error_log()` here, in addition to `WC_Stripe_Logger::error( $message );` in the following line?

Reply:
Yes. I want to make sure that we log _something_ even if logging is disabled at the plugin level -- someone needs to know about this no matter what. And I checked `WC_Stripe_Logger::error()` won't log if plugin-level logging is not enabled.